File: lib/scrapers/base_scraper.dart
import '../errors/scrape_exception.dart';
import '../models/filter_settings.dart';
import '../models/media_item.dart';

abstract class BaseScraper {
  bool canHandle(String url);

  Future<List<MediaItem>> extractMedia(String url, {FilterSettings? filters});
}

BaseScraper selectScraper(String url, List<BaseScraper> scrapers) {
  for (final scraper in scrapers) {
    if (scraper.canHandle(url)) return scraper;
  }

  throw ScrapeException('Нет парсера для URL', source: url);
}

File: app/src/main/java/io/github/rustnomicon/sourceinstaller/scrape/BaseScraper.kt
package io.github.rustnomicon.sourceinstaller.scrape

import io.github.rustnomicon.sourceinstaller.data.model.FilterSettings
import io.github.rustnomicon.sourceinstaller.data.model.MediaItem
import io.github.rustnomicon.sourceinstaller.errors.ScrapeException

interface BaseScraper {
    fun canHandle(url: String): Boolean

    suspend fun extractMedia(url: String, filters: FilterSettings = FilterSettings()): List<MediaItem>
}

fun selectScraper(url: String, scrapers: List<BaseScraper>): BaseScraper {
    for (scraper in scrapers) {
        if (scraper.canHandle(url)) return scraper
    }
    throw ScrapeException("Нет парсера для URL", source = url)
}
